pkg/cloudprovider/servicecontroller/servicecontroller.go

```
@@ -0,0 +1,468 @@
+		}
+		if exists && len(service.Spec.PublicIPs) == 0 {
+			// The load balancer exists, but we apparently don't know about its public
+			// IPs, so just delete it and recreate it to get back to a sane state.
+			// TODO: Ideally the cloud provider interface would return the IP for us.
+			glog.Infof("Deleting old LB for service with no public IPs %s", namespacedName)
+			if err := s.ensureLBDeleted(service); err != nil {
+				return err, retryable
+			}
+		} else if exists {
+			// TODO: Better handle updates for non-cached services, this is optimistic.
+			glog.Infof("LB already exists for service %s", namespacedName)
+			return nil, notRetryable
+		}
+	}
+
+	if !service.Spec.CreateExternalLoadBalancer {
+		glog.Infof("Not creating LB for service %s that doesn't want one.", namespacedName)
+		return nil, notRetryable
+	}
+
+	glog.V(2).Infof("Creating LB for service %s", namespacedName)
+
+	// The load balancer doesn't exist yet, so create it.
+	publicIPstring := fmt.Sprint(service.Spec.PublicIPs)
+	err := s.createExternalLoadBalancer(service)
+	if err != nil {
+		return fmt.Errorf("failed to create external load balancer for service %s: %v", namespacedName, err), retryable
+	}
+
+	if publicIPstring == fmt.Sprint(service.Spec.PublicIPs) {
+		glog.Infof("Not persisting unchanged service to registry.")
+		return nil, notRetryable
+	}
+
+	// If creating the load balancer succeeded, persist the updated service.
+	if err = s.persistUpdate(service); err != nil {
+		return fmt.Errorf("Failed to persist updated publicIPs to apiserver, even after retries. Giving up: %v", err), notRetryable
+	}
+	return nil, notRetryable
+}
+
+func (s *ServiceController) persistUpdate(service *api.Service) error {
+	var err error
+	for i := 0; i < clientRetryCount; i++ {
+		_, err = s.kubeClient.Services(service.Namespace).Update(service)
+		if err == nil {
+			return nil
+		}
+		if errors.IsNotFound(err) {
+			glog.Infof("Not persisting update to service that no longer exists: %v", err)
```

You're masking the error here again buy just logging it an returning success. You need to return an error so that the caller can reget and retry to update against the new version.